### PR TITLE
Deprecate adaptor factories.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR deprecates adaptor factory functions, per issue #1616.

After some deprecation cycle, these functions can be removed in a later release.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by vyasr — DISMISSED** [review]
I also see a `make_logging_adaptor` and (perhaps missed because it doesn't have `resource` in the name?).

**Comment by bdice** [comment]
> I also see a `make_logging_adaptor` and (perhaps missed because it doesn't have `resource` in the name?).

Fixed! Good catch.

**Comment by harrism** [comment]
@bdice are there any calls to these functions anywhere in RAPIDS? If so, we should update those calls to use the ctors first before merging this deprecation.

**Comment by bdice** [comment]
Yes, there are uses of these functions. They should be easy to fix. I have it on my list to tackle those.

**Comment by harrism** [comment]
OK, setting DO NOT MERGE label until this is done.

File: include/rmm/mr/device/limiting_resource_adaptor.hpp
Comment (wence-) [inline include/rmm/mr/device/limiting_resource_adaptor.hpp:193]: *nit*: Should we mention the version in which these were deprecated? (as per the docstring note we suggest for deprecations in cudf-python)

File: include/rmm/mr/device/polymorphic_allocator.hpp
Comment (wence-) [inline include/rmm/mr/device/polymorphic_allocator.hpp:307]: *issue*: Please update the docstring of `stream_allocator_adaptor` to use the preferred approach in its example.

**Review by wence- — CHANGES_REQUESTED** [review]
I think we also need the same deprecation on `make_prefetch_adaptor`.

Looks like `tests/logger_tests.cpp` and `tests/mr/device/stream_allocator_adaptor_tests.cpp` use these now deprecated functions.

The README file has an example using the `stream_allocator_adaptor` that also wants adapting for to the preferred constructor-based usage.

**Comment by harrism** [comment]
> I think we also need the same deprecation on make_prefetch_adaptor.

**Comment by bdice** [comment]
@wence- Thanks for the review. I am going to put this PR on a brief pause until I can remove the uses of `make_*_adaptor` elsewhere in RAPIDS, as Mark requested. This is targeting 24.10 so there's no rush for 24.08.

**Comment by harrism** [comment]
@bdice https://github.com/rapidsai/cugraph/pull/4572

I couldn't find any uses in other RAPIDS repos.

File: include/rmm/mr/device/logging_resource_adaptor.hpp
Comment (harrism) [inline include/rmm/mr/device/logging_resource_adaptor.hpp:310]: Theirs also the `friend` member function in the adaptor class above.

File: include/rmm/mr/device/polymorphic_allocator.hpp
Comment (harrism) [reply include/rmm/mr/device/polymorphic_allocator.hpp:307]: Yes, line 169 above. Also, the same example exists in RMM's `README.md`, please update there as well.

**Review by harrism — CHANGES_REQUESTED** [review]
Thanks for doing this.

Also need to update `logger_tests.cpp`, `adaptor_tests.cpp`, and `stream_allocator_adaptor_tests.cpp` to use the ctors and drop the factories.

File: include/rmm/mr/device/limiting_resource_adaptor.hpp
Comment (bdice) [reply include/rmm/mr/device/limiting_resource_adaptor.hpp]: Done.

File: include/rmm/mr/device/polymorphic_allocator.hpp
Comment (bdice) [reply include/rmm/mr/device/polymorphic_allocator.hpp:308]: Done.

File: include/rmm/mr/device/logging_resource_adaptor.hpp
Comment (bdice) [reply include/rmm/mr/device/logging_resource_adaptor.hpp:313]: Done.

**Comment by bdice** [comment]
@harrism @wence- This is ready for review. The two PRs to cudf and cugraph with changes to avoid the adaptor factories are now merged.

**Review by harrism — APPROVED** [review]
Thanks @bdice!

**Review by pmattione-nvidia — APPROVED** [review]

**Pull request merged**
